lib/semantics/check-do.cc

```diff
@@ -647,20 +643,124 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
+      .Attach(GetNodePosition(construct), "The construct that was left"_en_US);
+}
+
+static const parser::DoConstruct *MaybeGetDoConstruct(
+    const ConstructNode &construct) {
+  if (auto const *doNode{
+          std::get_if<const parser::DoConstruct *>(&construct)}) {
+    return *doNode;
+  }
+  return nullptr;
+}
+
+static bool ConstructIsDoConcurrent(const ConstructNode &construct) {
+  parser::DoConstruct const *doConstruct{MaybeGetDoConstruct(construct)};
+  return doConstruct && doConstruct->IsDoConcurrent();
+}
+
+// Check that CYCLE and EXIT statements do not cause flow of control to
+// leave DO CONCURRENT, CRITICAL, or CHANGE TEAM constructs.
+void DoChecker::CheckForBadLeave(
+    StmtType stmtType, const ConstructNode &construct) const {
+  std::visit(
+      common::visitors{
+          [&](const parser::DoConstruct *doConstructPtr) {
+            if (doConstructPtr->IsDoConcurrent()) {
+              // C1135 and C1167
+              SayBadLeave(stmtType, "DO CONCURRENT", construct);
+            }
+          },
+          [&](const parser::CriticalConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CRITICAL", construct);
+          },
+          [&](const parser::ChangeTeamConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CHANGE TEAM", construct);
+          },
+          [](const auto *) {},
+      },
+      construct);
+}
+
+static bool isExit(StmtType stmtType) { return stmtType == StmtType::EXIT; }
+
+static bool StmtMatchesConstruct(const parser::Name *stmtName,
+    StmtType stmtType, const parser::Name *constructName,
+    const ConstructNode &construct) {
+  bool inDoConstruct{MaybeGetDoConstruct(construct) != nullptr};
+  if (stmtName == nullptr) {
+    return inDoConstruct;  // Unlabeled statements match all DO constructs
+  } else if (constructName && constructName->source == stmtName->source) {
+    return isExit(stmtType) || inDoConstruct;
+  } else {
+    return false;
   }
 }
 
-// C1166
-void DoChecker::Enter(const parser::ExitStmt &) {
-  if (!context_.InsideDoConstruct()) {
-    context_.Say(
-        *context_.location(), "EXIT must be within a DO construct"_err_en_US);
+// C1167 Can't EXIT from a DO CONCURRENT
+void DoChecker::CheckDoConcurrentExit(
+    StmtType stmtType, const ConstructNode &construct) const {
+  if (isExit(stmtType) && ConstructIsDoConcurrent(construct)) {
+    SayBadLeave(StmtType::EXIT, "DO CONCURRENT", construct);
   }
 }
 
+// Check nesting violations for a CYCLE or EXIT statement.  Loop up the nesting
+// levels looking for a construct that matches the CYCLE or EXIT statment.  At
+// every construct, check for a violation.  If we find a match without finding
+// a violation, the check is complete.
+void DoChecker::CheckNesting(
+    StmtType stmtType, const parser::Name *stmtName) const {
+  const ConstructStack &stack{context_.constructStack()};
+  for (auto rIter{stack.crbegin()}; rIter != stack.crend(); ++rIter) {
```

Comment:
I understand that you're traversing `stack` from its most-recently-pushed entries to its oldest entries.  It would be less surprising if the loop used a decremented iterator and ran from `end()` back to `begin()` rather than using the rarely-used reversed iterator form `crbegin()`.  That is a facility that makes it easy to pass a reversed view of a container to an "algorithm" like `find_if()`, but one typically doesn't run into it in explicit looping situations.

```
for (auto iter{stack.cend()}; iter-- != stack.cbegin(); ) { ... }
```

Reply:
Cool!  I'll make this change.  Thanks, Peter.